engines/mohawk/myst_areas.cpp

```diff
@@ -184,16 +184,13 @@ MystAreaVideo::MystAreaVideo(MohawkEngine_Myst *vm, ResourceType type, Common::S
 		MystAreaAction(vm, type, rlstStream, parent) {
 	char c = 0;
 
-	do {
-		c = rlstStream->readByte();
+	while ((c = rlstStream->readByte()) != 0) {
 		_videoFile += c;
-	} while (c);
-
-	rlstStream->skip(_videoFile.size() & 1);
+	}
 
-	// Trim method does not remove extra trailing nulls
-	while (_videoFile.size() != 0 && _videoFile.lastChar() == 0)
-		_videoFile.deleteLastChar();
+	if ((_videoFile.size() & 1) == 0) {
```

Comment:
This looks wrong. Shouldnt it be != 0 ?

Reply:
That looks correct to me.

The way I understand that code, it looks like the stream always uses even number of bytes to store the string, so if we read an odd number of bytes, we need to read one my byte.

That's at least what the old code was doing since with the old `Common::String` code, adding the 0 would also increment the string size. So `_videoSize.size()` was giving us the number of bytes read.

With the new `Common::String` code, trying to add a 0 to the string does nothing (and in particular the size is not incremented). This is why this code broke. Now instead of checking is the size is odd, we need to check if it is even, because it does not count the additional terminating 0 (whether we try to add the 0 to the string or not). And this is what this new code is doing.

Reply:
That's the way I understood it too. My idea was to still read the same bytes from the file, but not to store the superfluous terminator in the resulting String. That way it doesn't matter whether or not String allows you to add extra terminators. But since that means _videoFile.size() is one less than it used to be, I had to invert the logic for when to skip a byte.

The idea behind the very untested Myst III patch above is also to not store the extra terminator. Myst III does work fine at the moment, but will break if the change to the String class is ever reverted.

I don't know where the Stark engine is impacted. I did manage to trigger the case where it adds 0 to the string by just waiting around in the dream at the beginning of the game. But I wasn't able to reproduce it when I tried running in a debugger.

Reply:
Then LGTM from me, feel free to merge